enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+#### Kube Controller Manager and Kube API Server
+
+Configuration of these components is managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively. Both operators use [`configobserver`](https://github.com/openshift/library-go/blob/16d6830d0b80dc2a3315207116d009ed2dd4cebf/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L140) module from `library-go` where we set what cloud provider should be used for the given platform. When a platform is being switched to `CCM`, we [must not](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) set `cloudProvider` value there and leave it empty.
+
+Next step is to bump `library-go` for the operators.
+
+##### Pre-release / Development
+
+In order to achieve desired cluster configuration for cloud components during development, we need to add an annotation on the `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11).
+
+Proposed annotation: `infrastructure.openshift.io/external-cloud-provider`.
+
+The annotation will determine the architectural change in the cluster configuration. If this annotation would be found, other components, such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will set their flags accordingly, to allow hosting external `CCM` for any platform.
+
+The annotation will be a temporary measure to simplify the operator development allowing multiple components to merge implementations of this feature independently before switching on the feature by changing the value of the annotation. Once all providers are moved out-of-tree, the annotation will be safe to remove.
+
+To ensure a smooth transition from the `in-tree` selection, `library-go` [implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) will include an additional argument for the `cloudProviderObserver` structure, which will allow to preserve initial configuration strategy, until the new `observeExternal` flag will be set to `false`:
+
+```go
+type cloudProviderObserver struct {
+    targetNamespaceName     string
+    cloudProviderNamePath   []string
+    cloudProviderConfigPath []string
+    observeExternal         bool
+}
+```
+
+The `observeExternal` field will determine, when the annotation is present on the `Infrastructure` resource, it will be taken into account while establishing needed flags for a Kubernetes binaries. It will make the observer type to the `cloud-config` file, and set flags as was described in the `kubelet`, `kube-apiserver` and `kube-controller-manager` sections.
```

Comment:
is `observeExternal` state of the controller? Is it an argument for the observer to behave differently? Why would we need that? Why not hardcoded (as @soltysh suggests) ?

Reply:
Dropping this info, it is an implementational detail. I still expect an annotation on the Infra resource to be used in development.